File: go/ppc/match.go
package capitan

import (
	"fmt"
	"regexp"
	"strings"
)

type Hole struct {
	Pattern  string
	Variable string
	Offset   int
	Kind     string
}

func (h Hole) String() string {
	return fmt.Sprintf("Hole{Pattern: %q, Variable: %q, Offset: %d, Kind: %q}", h.Pattern, h.Variable, h.Offset, h.Kind)
}

type Constant struct {
	Value string
}

func (c Constant) String() string {
	return fmt.Sprintf("Constant{Value: %q}", c.Value)
}

type Match interface {
	Hole | Constant
}

func ParseTemplate(input string) ([]Match, error) {
	var matches []Match
	offset := 0

	for len(input) > 0 {
		hole, rest, err := parseHole(input)
		if err == nil {
			hole.Offset = offset
			matches = append(matches, hole)
			input = rest
			offset += len(hole.Pattern)
			continue
		}

		constant, rest, err := parseConstant(input)
		if err == nil {
			matches = append(matches, constant)
			input = rest
			offset += len(constant.Value)
			continue
		}

		return nil, fmt.Errorf("failed to parse template at offset %d: %q", offset, input)
	}

	return matches, nil
}

func parseHole(input string) (Hole, string, error) {
	hole, err := parseHolePattern(input, "", "")
	if err == nil {
		return hole, input[len(hole.Pattern):], nil
	}

	hole, rest, err := parseHolePattern(input, "[", "]")
	if err == nil {
		return hole, rest, nil
	}

	hole, rest, err = parseHolePattern(input, "{", "}")
	if err == nil {
		return hole, rest, nil
	}

	hole, rest, err = parseHolePattern(input, "(", ")")
	if err == nil {
		return hole, rest, nil
	}

	hole, rest, err = parseHolePattern(input, "<", ">")
	if err == nil {
		return hole, rest, nil
	}

	hole, rest, err = parseRegexHole(input)
	if err == nil {
		return hole, rest, nil
	}

	return Hole{}, input, fmt.Errorf("failed to parse hole")
}

func parseHolePattern(input, left, right string) (Hole, string, error) {
	leftPattern := regexp.MustCompile(fmt.Sprintf(`^%s([a-zA-Z_']+)%s`, left, right))
	matches := leftPattern.FindStringSubmatch(input)
	if len(matches) == 0 {
		return Hole{}, input, fmt.Errorf("failed to parse hole pattern")
	}

	hole := Hole{
		Pattern:  matches[0],
		Variable: matches[1],
		Kind:     "value",
	}

	return hole, input[len(hole.Pattern):], nil
}

func parseRegexHole(input string) (Hole, string, error) {
	pattern := regexp.MustCompile(`^\$([a-zA-Z_']+)~(.+)\$`)
	matches := pattern.FindStringSubmatch(input)
	if len(matches) == 0 {
		return Hole{}, input, fmt.Errorf("failed to parse regex hole")
	}

	hole := Hole{
		Pattern:  matches[0],
		Variable: matches[1],
		Kind:     "value",
	}

	return hole, input[len(hole.Pattern):], nil
}

func parseConstant(input string) (Constant, string, error) {
	pattern := regexp.MustCompile(`^([^$]+)`)
	matches := pattern.FindStringSubmatch(input)
	if len(matches) == 0 {
		return Constant{}, input, fmt.Errorf("failed to parse constant")
	}

	constant := Constant{
		Value: matches[1],
	}

	return constant, input[len(constant.Value):], nil
}

File: python/ppc/match.py
import re

class Hole:
    def __init__(self, pattern, variable, offset, kind="value"):
        self.pattern = pattern
        self.variable = variable
        self.offset = offset
        self.kind = kind

    def __repr__(self):
        return f"Hole(pattern='{self.pattern}', variable='{self.variable}', offset={self.offset}, kind='{self.kind}')"

class Constant:
    def __init__(self, value):
        self.value = value

    def __repr__(self):
        return f"Constant(value='{self.value}')"

def parse_template(input_str):
    matches = []
    offset = 0

    while input_str:
        for pattern, kind in [
            (r"^(\[\[.*?\]\])", "value"),
            (r"^(\{\{.*?\}\})", "value"),
            (r"^(\(.*?\))", "value"),
            (r"^(\<.*?\>)", "value"),
            (r"^\$([a-zA-Z_']+)~(.+)\$", "value"),
        ]:
            match = re.match(pattern, input_str)
            if match:
                pattern_str, variable = match.groups()[:2]
                hole = Hole(pattern_str, variable, offset, kind)
                matches.append(hole)
                offset += len(pattern_str)
                input_str = input_str[len(pattern_str):]
                break

        else:
            constant_match = re.match(r"^([^$]+)", input_str)
            if constant_match:
                constant_value = constant_match.group(1)
                constant = Constant(constant_value)
                matches.append(constant)
                offset += len(constant_value)
                input_str = input_str[len(constant_value):]
            else:
                break

    return matches